src/Helpers/DataResolver.php

```diff
@@ -0,0 +1,159 @@
+     * @param array $columns
+     *
+     * @return void
+     * @throws Exception
+     */
+    protected function cannotIdentifyException($component, $columns = [])
+    {
+        if (!empty($columns)) {
+            throw new Exception(
+                sprintf(
+                    'Cannot identify, "%s" from class "%s"',
+                    implode('.', $columns),
+                    get_class($component)
+                )
+            );
+        }
+    }
+
+    /**
+     * Resolves a Single field in the database.
+     * @return mixed
+     * @throws Exception
+     */
+    protected function resolveField()
+    {
+        $this->cannotIdentifyException($this->component, $this->columns);
+        if ($this->columnName) {
+            $method = $this->columnName;
+            if (!method_exists($this->component, $method)) {
+                throw new Exception(
+                    sprintf('Method, "%s" not found on "%s"', $method, get_class($this->component))
+                );
+            }
+            return $this->component->$method();
+        }
+        return $this->component->getValue();
+    }
+
+    /**
+     * Resolves a DataObject value
+     * @return mixed
+     * @throws Exception
+     */
+    protected function resolveDataObject()
+    {
+        if (empty($this->columnName)) {
+            return $this->component->toMap();
+        }
+        // Inspect component for element $relation
+        if ($this->component->hasMethod($this->columnName)) {
+            $relation = $this->columnName;
+            return self::identify($this->component->$relation(), $this->columns);
+        }
+        // Inspect component has attribute
+        if ($this->component->hasField($this->columnName)) {
+            $data = $this->component->{$this->columnName};
+            $dbObject = $this->component->dbObject($this->columnName);
+            if ($dbObject) {
+                $dbObject->setValue($data);
+                return self::identify($dbObject, $this->columns);
+            }
+            return $data;
+        }
+        $this->cannotIdentifyException($this->component, [$this->columnName]);
+    }
+
+    /**
+     * Resolves an ArrayData value
+     * @return mixed
+     * @throws Exception
+     */
+    public function resolveArrayData()
+    {
+        if (empty($this->columnName)) {
+            return $this->component->toMap();
+        }
+        // Inspect component has attribute
+        if ($this->component->hasField($this->columnName) && empty($this->columns)) {
+            return $this->component->{$this->columnName};
+        }
+        $this->cannotIdentifyException($this->component, array_merge([$this->columnName], $this->columns));
+    }
+
+    /**
+     * Resolves a DataList values
+     * @return array|mixed
+     * @throws Exception
+     */
+    public function resolveList()
+    {
+        if (empty($this->columnName)) {
+            return $this->component->toNestedArray();
```

Comment:
How memory intensive is this?

Reply:
I am not sure, `toNestedArray()` is the most simple and generic method I like

Reply:
No stress, just checking, as it might be worth looking in to :)
The indexing and configuring are already heavy anyway, so it's better than the old system either way, but the more power useage we can shave off, the better :)